kernel/src/common/deferred_call_mux.rs

```diff
@@ -0,0 +1,207 @@
+/// is determined by the `clients`-array passed in with the constructor.
+pub struct DeferredCallMux {
+    deferred_call: DeferredCall<usize>,
+    clients: &'static [(Cell<bool>, OptionalCell<&'static DeferredCallMuxClient>)],
+    handle_counter: Cell<usize>,
+}
+
+impl DeferredCallMux {
+    /// Construct a new deferred call
+    ///
+    /// This needs to be registered with the `set_global_mux` function immediately
+    /// afterwards, and should not be changed anymore. Only the globally registered
+    /// Mux will receive calls from the underlying deferred call implementation.
+    ///
+    /// The `clients` array can be initialized with any value. The recommended
+    /// values are (where n is the number of possible clients):
+    ///
+    /// `[(Cell::new(false), OptionalCell::empty()); n]`
+    pub unsafe fn new(
+        clients: &'static [(Cell<bool>, OptionalCell<&'static DeferredCallMuxClient>)],
+    ) -> DeferredCallMux {
+        DeferredCallMux {
+            deferred_call: DeferredCall::new(DEFERRED_CALL_MUX_TASK),
+            clients: clients,
+            handle_counter: Cell::new(0),
+        }
+    }
+
+    /// Schedule a deferred call to be called
+    ///
+    /// The handle addresses the client that will be called.
+    ///
+    /// If no client for the handle is found (it was unregistered), this
+    /// returns `None`. If a call is already scheduled, it returns
+    /// `Some(false)`.
+    pub fn set(&self, handle: DeferredCallHandle) -> Option<bool> {
+        let DeferredCallHandle(client_pos) = handle;
+        let client = &self.clients[client_pos];
+
+        if let (call_set, true) = (&client.0, client.1.is_some()) {
+            if call_set.get() {
+                // Already set
+                Some(false)
+            } else {
+                call_set.set(true);
+                self.deferred_call.set();
+                Some(true)
+            }
+        } else {
+            None
+        }
+    }
+
+    /// Register a new client
+    ///
+    /// On success, a `Some(handle)` will be returned. This handle is later
+    /// required to schedule a deferred call.
+    pub fn register(
+        &self,
+        mux_client: &'static DeferredCallMuxClient,
+    ) -> Option<DeferredCallHandle> {
+        let current_counter = self.handle_counter.get();
+
+        if current_counter < self.clients.len() {
+            let client = &self.clients[current_counter];
+            client.0.set(false);
+            client.1.set(mux_client);
+
+            self.handle_counter.set(current_counter + 1);
+
+            Some(DeferredCallHandle(current_counter))
+        } else {
+            None
+        }
+    }
+
+    /// Call all registered and to-be-scheduled deferred calls
+    ///
+    /// This function needs to be called by the underlying deferred call implementation.
+    /// It may be called without holding the `DeferredCallMux` reference through
+    /// `call_global_mux`.
+    pub(self) fn call(&self) {
```

Comment:
why is `pub(self)` needed?

Reply:
The `call`-method should really only be called from inside this module, via the `call_global_mux` function. I don't see any case where further exposing this method would be helpful. That's why it's only exposed to the current module.